recipe-server/normandy/recipes/models.py

```diff
@@ -517,19 +550,55 @@ def recipes_used_by(self):
     def __str__(self):
         return self.name
 
+    def canonical_json(self):
+        # Avoid circular import
+        from normandy.recipes.api.v1.serializers import ActionSerializer
+        data = ActionSerializer(self).data
+        return CanonicalJSONRenderer().render(data)
+
     def get_absolute_url(self):
         return reverse('action-detail', args=[self.name])
 
     def compute_implementation_hash(self):
         return hashlib.sha1(self.implementation.encode()).hexdigest()
 
+    def update_signature(self):
+        try:
+            autographer = Autographer()
+        except ImproperlyConfigured:
+            self.signature = None
+            return
+
+        logger.info(
+            f'Requesting signatures for action named {self.name} from Autograph',
+            extra={'code': INFO_REQUESTING_ACTION_SIGNATURES, 'action_names': [self.name]}
+        )
+
+        signature_data = autographer.sign_data([self.canonical_json()])[0]
+        signature = Signature(**signature_data)
+        signature.save()
+        self.signature = signature
+
+    @transaction.atomic
     def save(self, *args, **kwargs):
-        # Save first so the upload is available.
-        super().save(*args, **kwargs)
+        if self.is_dirty(check_relationship=True):
+            dirty_fields = self.get_dirty_fields(check_relationship=True)
+            dirty_field_names = list(dirty_fields.keys())
+
+            if (len(dirty_field_names) > 1 and 'signature' in dirty_field_names
+                    and self.signature is not None):
+                # Setting the signature while also changing something else is probably
+                # going to make the signature immediately invalid. Don't allow it.
+                raise ValidationError('Signatures must change alone')
```

Comment:
I resisted doing this because it is only the second time we've done this, and it was already pretty hard to keep everything straight. I tried to limit the scope of the change to get it out faster and avoid breaking existing code.